tools/autograd/load_derivatives.py

```diff
@@ -87,11 +89,175 @@ def create_derivative(f: NativeFunction, formula: str, var_names: Tuple[str, ...
+        elif formula == "auto_linear":
+            if len(forward_derivatives) > 1:
+                raise RuntimeError(f"Derivative definition of {defn_name} in derivatives.yaml defines the "
+                                   "forward definition of gradient as linear but this only works "
+                                   "for functions with a single differentiable output.")
+            # This transformation is based on the observation that linear functions can be written as:
+            #   y = f(x) = A * x
+            # For some matrix A and the Jacobian of the function f is also A.
+            # So doing J * v = A * v = f(v).
+            # Hence to do the jvp, we simply need to evaluate the function at the point v instead of x.
+            # We do this by calling the forward again by replacing any occurrence of the differentiable
+            # input "foo" by it's tangent "foo_t".
+            # Note that multiple inputs are not a problem as long as the function is truly linear wrt to
+            # the vector where all the differentiable inputs are stacked.
+
+            diff_arg_names = [arg.name for arg in args_with_derivatives]
+            assert len(diff_arg_names) > 0
+
+            # Do replacement of input variables
+            new_args = []
+            for arg_name in all_arg_names:
+                if arg_name in diff_arg_names:
+                    arg_name = arg_name + "_t"
+                new_args.append(arg_name)
+
+            # Call into the forward again. We need two cases here to handle both Tensor methods and at:: functions.
+            if Variant.function in f.variants:
+                fw_formula = "at::{}({})".format(defn_name, ", ".join(new_args))
+            else:
+                assert f.func.kind() is not SchemaKind.inplace
+                assert Variant.method in f.variants
+                fw_formula = "{}.{}({})".format(new_args[0], defn_name, ", ".join(new_args[1:]))
+
+            # All of the input tangents are always used so all of them are required here.
+            required_inputs_tangent = tuple(diff_arg_names)
+            formula = fw_formula
+
+        # At this point, the formula is final and is not modified anymore.
+
+        # During forward formula, we use the primal instead of the input Tensors.
+        # This call inspects the formula to find for which input's primal are used.
+        required_inputs_primal = find_required_inputs(formula, "_p")
+
+        updated_derivatives.append(ForwardDerivative(
+            formula=formula,
+            var_name=defn.var_name,
+            var_type=defn.var_type,
+            required_inputs_fw_grad=required_inputs_tangent,
+            required_inputs_primal=required_inputs_primal))
+
+    return updated_derivatives
+
+def is_forward_derivative_definition(all_arg_names: List[str], names: Tuple[str, ...]) -> bool:
+    if len(names) > 1:
+        # Forward definition are always for a single output at a time
+        return False
```

Comment:
What happens if someone writes "result0, result1: blah"? Will the dev see a nice error message or will they be confused?

Reply:
The first line of create derivatives will give a nice error: assert len(names) == 1, "Forward derivatives can define gradients for only one output at a time"